Primary sites: document new in order streaming configuration (#283)
foxglove/docs: docs/10-primary-sites/1-self-hosting/6-inbox-listener-configuration.md

> ---

Reviewer: I think this page would work well as the second page in the section — after "installation" but before "manage data", since the latter describes importing data.

title: Inbox listener configuration
description: Configure the inbox listener in your self-hosted Primary Site.
---

### Configure the inbox listener

The inbox listener consumes inputs in the inbox bucket and writes the split outputs into the lake bucket. One of its responsibilities is to ensure that the messages in the lake are stored in log-time order, even if the inbox file contained out-of-order messages.

It ensures log-time order using one of two strategies:

> 1. For indexed MCAP inputs, it uses a buffered in-order reader. When an MCAP file is detected, its [chunk indexes](https://mcap.dev/spec#chunk-index-op0x08) are analyzed to determine the buffer size required to read the file in order. If this buffer size is greater than a configured limit, it switches to the fallback strategy.

Reviewer 2: I think this is a higher level of detail than we need to be documenting? Why does a deployer care about a buffered in-order reader? This is an implementation detail.

Author: I was trying to leave in enough detail to justify why there is a buffer which has a size that you can adjust. Do you have any thoughts on how we could be less specific here?

Reviewer 2: I guess after further review I don't even find myself swayed on why this is a config option at all (or rather why it is so specific). To me this reads a bit like "process memory limit"? Or something potentially more generic like that. I think the point I am going after is that it feels like a knob that should not exist in its proposed form.

Another thing I feel like this page lacks is the call-out that in most cases you don't need to change any configuration? Like why are we making this harder for the user to deploy?

If I had to copy-edit this page I would be writing something like:

```
The inbox listener processes recordings from the inbox bucket to the lake bucket for indexing, storage, and access.

It uses a combination of compute, memory, and disk resources and can be scaled horizontally (see Autoscaling). Depending on the size and structure of your recordings, you may need to adjust a few configuration options to ensure files are processed.

`MAX_SCRATCH_CACHE_BUFFER_BYTES`: The maximum reader buffer size that will be used for storing recording data in-memory while processing files. During processing, some file data is read into memory before being written to the lake. This setting controls the maximum scratch buffer. If unset, there is no limit. If set and the limit is exceeded, a local or remote disk is used instead of memory.

`LOCAL_SCRATCH_ROOT`: The directory on the local filesystem that the inbox listener may use for scratch storage. If unconfigured, the inbox listener will use the lake bucket as a scratch storage instead, which is usually slower.

`LOCAL_SCRATCH_ROOT_CAPACITY_BYTES`: The available storage capacity of the `LOCAL_SCRATCH_ROOT` in bytes. If an input requires more capacity, a warning is logged and it will use the lake bucket as scratch storage instead, which is usually slower.

> 2. It uses a fallback strategy for unindexed MCAP and all ROS 1 BAG inputs. The input is split into scratch storage and then each output is re-ordered as a post-processing step. It will use the local disk if configured with a scratch directory, otherwise it will use the lake bucket as scratch storage.

Reviewer: ```suggestion
2. It uses a fallback strategy for unindexed MCAP and all ROS 1 BAG inputs. The input is split into scratch storage and then each output is re-ordered as a post-processing step. It will use the local disk if configured with a scratch directory; otherwise, it will use the lake bucket as scratch storage.
```

| --------------------------------- | ---------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `reader_chunk_buffer_limit_bytes` | 1073741824 | The maximum reader buffer size that will be used for reading indexed MCAP. If an input contains highly out-of-order messages it may exceed this limit. In this case a warning is logged and the inbox listener switches to the fallback strategy. |
| `local_storage_root`              |            | The directory on the local filesystem that the inbox listener may use for scratch storage. If unconfigured, the inbox listener will use the lake bucket as a scratch storage instead, which is usually slower.                                    |
| `local_storage_capacity_bytes`    | 0          | The available storage capacity of the `local_storage_root` in bytes. If an input requires more capacity, a warning is logged and it will use the lake bucket as scratch storage instead, which is usually slower.                                 |

> If your primary site is expected to process unindexed MCAP or ROS 1 BAG files, consider configuring your inbox listener deployment with [ephemeral storage](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir). A good rule of thumb is to allocate at least three times as much capacity as your largest expected input file size.

Reviewer 2: ```suggestion
If your primary site is expected to process unindexed MCAP or ROS 1 BAG files, consider configuring your inbox listener deployment with [ephemeral storage](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir). We recommend allocating at least three times as much capacity as your largest expected input file size.
```
